pallets/settlement/src/lib.rs

```diff
@@ -585,19 +649,43 @@ decl_module! {
         ///
         /// # Arguments
         /// * `instruction_id` - Instruction id to authorize.
-        /// * `portfolios` - Portfolios that the sender controls and wants to authorize for this instruction
+        /// * `portfolios` - Portfolios that the sender controls and wants to authorize for this instruction.
         #[weight = weight_for::weight_for_authorize_instruction::<T>()
-            + weight_for::weight_for_transfer::<T>() // Maximum weight for `execute_instruction()`
+            + weight_for::weight_for_transfer::<T>() // Maximum weight for `execute_instruction()`.
         ]
         pub fn authorize_instruction(origin, instruction_id: u64, portfolios: Vec<PortfolioId>) -> DispatchResultWithPostInfo {
             let sender = ensure_signed(origin)?;
             let did = Context::current_identity_or::<Identity<T>>(&sender)?;
             let portfolios_set = portfolios.into_iter().collect::<BTreeSet<_>>();
 
-            // Authorize the instruction
+            // Authorize the instruction.
+            Self::unsafe_authorize_instruction(did, instruction_id, portfolios_set)?;
+
+            // Execute the instruction if conditions are met.
+            let auths_pending = Self::instruction_auths_pending(instruction_id);
+            let weight_for_instruction_execution = Self::is_instruction_executed(auths_pending, Self::instruction_details(instruction_id).settlement_type, instruction_id);
+
+            Ok(Some(weight_for::weight_for_authorize_instruction::<T>() + weight_for_instruction_execution).into())
+        }
+
+        /// Authorizes an existing confidential instruction.
+        ///
+        /// # Arguments
+        /// * `instruction_id` - Instruction id to authorize.
+        /// * `portfolios` - Portfolios that the sender controls and wants to authorize for this instruction.
+        #[weight = weight_for::weight_for_authorize_instruction::<T>()
+            + weight_for::weight_for_transfer::<T>() // Maximum weight for `execute_instruction()`
+        ]
+        pub fn authorize_confidential_instruction(origin, instruction_id: u64, data: MercatTxData, portfolios: Vec<PortfolioId>) -> DispatchResultWithPostInfo {
+            let sender = ensure_signed(origin)?;
+            let did = Context::current_identity_or::<Identity<T>>(&sender)?;
+            let portfolios_set = portfolios.into_iter().collect::<BTreeSet<_>>();
+
+            // Authorize the instruction.
             Self::unsafe_authorize_instruction(did, instruction_id, portfolios_set)?;
 
```

Comment:
Can't we just dedup this with the non-confidential function part?

Reply:
We can, if you are ok with receiving an option type in the `authorize_instruction` function. You can pass None and I will pass `Some(MercatTxData)`

Reply:
I am referring to 3 -4 lines that are the same in both dispatchable.

Reply:
The first 3-4 lines are checking the caller's identity, the common logic is in `unsafe_authorize_instruction()` which is reused. I also don't want to refactor the part that's calculating weights, since eventually we are going to have different weights for confidential transfers. So I'm siding with not touching this code for now.